newsrevenuehub/rev-engine
# DEV-2689: Fix autocomplete in donor address field

Please fill out each section even if it's just with "N/A"

#### Plan? (if this draft/incomplete indicate what you intend to do and how)

n/a

#### What's this PR do?

- Fixes a bug with autofilling a contribution form's address fields. It doesn't fix the country field; that will be done in DEV-2691.
- Migrates DDonorAddress to TypeScript, uses our base components, and adds Jest tests.

#### Why are we doing this? How does it help us?

- Improves UX by allowing autofill to work.
- Continues the code migration.

#### How should this be manually tested? Please include detailed step-by-step instructions.

**Updated based on problems seen in previous review apps**

Testing autofill:

1. Visit a contribution page.
2. Confirm the address fields match appearance of an existing deploy (e.g. there were no appearance changes accidentally introduced). One thing to double-check in particular is that focused fields pick up the page style color.
3. Confirm that all address fields are required (in this PR, they will block submission instead of showing an error message below them).
4. Use browser autofill to put in an address. It should fill out all fields except the Country field.
5. Make a test contribution.
6. Confirm that the address you entered appears in the Contributions section as well as Stripe.

Testing Google Maps autocomplete:

1. Visit a contribution page.
2. Enter an address into the Address field.
3. Confirm that all address fields are filled in except Country, but name and other fields are unaffected.
4. Fill in the rest of the fields and complete the test contribution.
5. Confirm that the address you entered appears in the Contributions section as well as Stripe.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

I think the change to the required fields behavior isn't major enough to require documentation, but that's just my opinion.

#### Have automated unit tests been added? If not, why?

Yes.

#### How should this change be communicated to end users?

See documentation question above.

#### Are there any smells or added technical debt to note?

No.

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

https://news-revenue-hub.atlassian.net/browse/DEV-2689

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No.

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

https://news-revenue-hub.atlassian.net/browse/DEV-2691
